# Objective
 The objective is to use linear regression to find the median value of owner-occupied homes in 1000 USD's.
 We will build a Machine learning model (i.e. Linear Regression) using tensorflow.keras
 
 Data : https://raw.githubusercontent.com/dphi-official/Datasets/master/Boston_Housing/Training_set_boston.csv

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/Boston_Housing/Training_set_boston.csv')
df.head()

CRIM    ZN  INDUS  CHAS     NOX     RM   AGE     DIS   RAD    TAX  \
0  15.02340   0.0  18.10   0.0  0.6140  5.304  97.3  2.1007  24.0  666.0   
1   0.62739   0.0   8.14   0.0  0.5380  5.834  56.5  4.4986   4.0  307.0   
2   0.03466  35.0   6.06   0.0  0.4379  6.031  23.3  6.6407   1.0  304.0   
3   7.05042   0.0  18.10   0.0  0.6140  6.103  85.1  2.0218  24.0  666.0   
4   0.72580   0.0   8.14   0.0  0.5380  5.727  69.5  3.7965   4.0  307.0   

   PTRATIO       B  LSTAT  MEDV  
0     20.2  349.48  24.91  12.0  
1     21.0  395.62   8.47  19.9  
2     16.9  362.25   7.83  19.4  
3     20.2    2.52  23.29  13.4  
4     21.0  390.95  11.28  18.2

### Separating input and output features

In [3]:
X = df.drop('MEDV', axis = 1)
y = df.MEDV

### Splitting dataset into train and test data

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 29)

In [6]:
# no of input features
n_features = X.shape[1]
n_features

13

### Training model

#### 1) Defining model

In [8]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from numpy.random import seed
import tensorflow

In [9]:
# defining model
model = Sequential()
model.add(Dense(10, activation = 'relu', input_shape = (n_features,)))
model.add(Dense(8, activation = 'relu'))
model.add(Dense(1))      
#Note that the visible layer of the network is defined by the “input_shape” argument on the first hidden layer.

#### 2) Compiling model

In [10]:
# import RMSprop optimimzer
from tensorflow.keras.optimizers import RMSprop
optimizer = RMSprop(0.01)                            # 0.01 is the learning rate

In [11]:
model.compile(loss = 'mean_squared_error', optimizer = optimizer)       # compling the model

#### 3) Fitting model

In [12]:
seed_value = 29
seed(seed_value)        # If you build the model with given parameters, set_random_seed will help you produce the same result on multiple execution

# Recommended by Keras -------------------------------------------------------------------------------------
# 1. Set `PYTHONHASHSEED` environment variable at a fixed value
import os
os.environ['PYTHONHASHSEED']=str(seed_value)

# 2. Set `python` built-in pseudo-random generator at a fixed value
import random
random.seed(seed_value)

# 3. Set `numpy` pseudo-random generator at a fixed value
import numpy as np
np.random.seed(seed_value)

# 4. Set the `tensorflow` pseudo-random generator at a fixed value
tensorflow.random.set_seed(seed_value) 


model.fit(X_train, y_train, epochs = 10, batch_size = 30, verbose = 1)

Epoch 1/10
11/11 [==============================] - 1s 3ms/step - loss: 850.7260
Epoch 2/10
11/11 [==============================] - 0s 2ms/step - loss: 238.9503
Epoch 3/10
11/11 [==============================] - 0s 2ms/step - loss: 200.1113
Epoch 4/10
11/11 [==============================] - 0s 2ms/step - loss: 146.1230
Epoch 5/10
11/11 [==============================] - 0s 1ms/step - loss: 269.7995
Epoch 6/10
11/11 [==============================] - 0s 1ms/step - loss: 110.7625
Epoch 7/10
11/11 [==============================] - 0s 1ms/step - loss: 218.2884
Epoch 8/10
11/11 [==============================] - 0s 1ms/step - loss: 136.6230
Epoch 9/10
11/11 [==============================] - 0s 2ms/step - loss: 168.5215
Epoch 10/10
11/11 [==============================] - 0s 1ms/step - loss: 140.7047


By setting verbose 0, 1 or 2 you just say how do you want to 'see' the training progress for each epoch.

verbose=0 will show you nothing (silent)

verbose=1 will show you an animated progress bar like this:

progres_bar

verbose=2 will just mention the number of epoch like this:

verbose = 2

#### 4) Evaluating model

In [13]:
model.evaluate(X_test, y_test)

3/3 [==============================] - 0s 0s/step - loss: 116.6633


116.66329956054688

The mean squared error we got here is 116.66. Now, what does it mean?

When you subtract the predicted values (of X_test data) from the acutal value (of X_test data), then square it and sum all the squares, and finally take a mean (i.e. average) of it, the result you will get is 116.66 in this case.

evaluate() does this task automatically. If you want to get the prediciton for X_test you can do model.predict(X_test)

### Hyperparameter tuning

#### 1) Learning rate

In [15]:
# defining model
model = Sequential()
model.add(Dense(10, activation = 'relu', input_shape = (n_features,)))
model.add(Dense(8, activation = 'relu'))
model.add(Dense(1))

# compiling model
optimizer = RMSprop(0.1)
model.compile(loss = 'mean_squared_error', optimizer = optimizer)

# fitting model
model.fit(X_train, y_train, epochs = 10, batch_size = 30, verbose = 1)

# evaluating model
print('The MSE value is:', model.evaluate(X_test, y_test))

Epoch 1/10
11/11 [==============================] - 0s 2ms/step - loss: 549237.5000
Epoch 2/10
11/11 [==============================] - 0s 2ms/step - loss: 1205.2281
Epoch 3/10
11/11 [==============================] - 0s 2ms/step - loss: 341.3297
Epoch 4/10
11/11 [==============================] - 0s 1ms/step - loss: 1116.0170
Epoch 5/10
11/11 [==============================] - 0s 2ms/step - loss: 5351.7485
Epoch 6/10
11/11 [==============================] - 0s 1ms/step - loss: 418.6449
Epoch 7/10
11/11 [==============================] - 0s 1ms/step - loss: 2980.2480
Epoch 8/10
11/11 [==============================] - 0s 1ms/step - loss: 66.5329
Epoch 9/10
11/11 [==============================] - 0s 1ms/step - loss: 79.2075
Epoch 10/10
3/3 [==============================] - 0s 0s/step - loss: 144.1426
The MSE value is: 144.14260864257812


In [32]:
# playing with different learning rates
learning_rate = 0.05
epochs = 10
optimizer = RMSprop(learning_rate)
model.compile(loss = 'mean_squared_error', optimizer = optimizer)
model.fit(X_train, y_train, epochs = epochs, batch_size = 30)
model.evaluate(X_test, y_test)

Epoch 1/10
11/11 [==============================] - 0s 2ms/step - loss: 85.4150
Epoch 2/10
11/11 [==============================] - 0s 3ms/step - loss: 85.4060
Epoch 3/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3506
Epoch 4/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3248
Epoch 5/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3334
Epoch 6/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3137
Epoch 7/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3816
Epoch 8/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3588
Epoch 9/10
11/11 [==============================] - 0s 2ms/step - loss: 85.3342
Epoch 10/10
3/3 [==============================] - 0s 6ms/step - loss: 93.9497


93.94967651367188

#### 2) Epochs

In [34]:
model = Sequential()
model.add(Dense(10, activation = 'relu', input_shape = (n_features,)))
model.add(Dense(8, activation = 'relu'))
model.add(Dense(1))

optimizer = RMSprop(0.01)

model.compile(loss = 'mean_squared_error', optimizer = optimizer)
model.fit(X_train, y_train, epochs = 100, batch_size = 30, verbose = 1)
model.evaluate(X_test, y_test)

Epoch 1/100
11/11 [==============================] - 0s 2ms/step - loss: 2917.3066
Epoch 2/100
11/11 [==============================] - 0s 2ms/step - loss: 359.4423
Epoch 3/100
11/11 [==============================] - 0s 2ms/step - loss: 143.9291
Epoch 4/100
11/11 [==============================] - 0s 2ms/step - loss: 69.2657
Epoch 5/100
11/11 [==============================] - 0s 2ms/step - loss: 92.6498
Epoch 6/100
11/11 [==============================] - 0s 2ms/step - loss: 58.6144
Epoch 7/100
11/11 [==============================] - 0s 1ms/step - loss: 79.3011
Epoch 8/100
11/11 [==============================] - 0s 3ms/step - loss: 71.3325
Epoch 9/100
11/11 [==============================] - 0s 2ms/step - loss: 65.3908
Epoch 10/100
11/11 [==============================] - 0s 2ms/step - loss: 71.6543
Epoch 11/100
11/11 [==============================] - 0s 1ms/step - loss: 71.8583
Epoch 12/100
11/11 [==============================] - 0s 2ms/step - loss: 68.7715
Epoch 13/100
11/11 [=

3/3 [==============================] - 0s 0s/step - loss: 23.8720


23.871976852416992

In [45]:
# Playing with epochs and learning rate
model = Sequential()
model.add(Dense(10, activation = 'relu', input_shape = (n_features,)))
model.add(Dense(8, activation = 'relu'))
model.add(Dense(1))
learning_rate = 0.01
epochs = 100
optimizer = RMSprop(learning_rate)

model.compile(loss = 'mean_squared_error', optimizer = optimizer)
model.fit(X_train, y_train, epochs = epochs, batch_size = 30)
model.evaluate(X_test, y_test)

Epoch 1/100
11/11 [==============================] - 0s 1ms/step - loss: 5083.3569
Epoch 2/100
11/11 [==============================] - 0s 2ms/step - loss: 168.6192
Epoch 3/100
11/11 [==============================] - 0s 943us/step - loss: 139.8480
Epoch 4/100
11/11 [==============================] - 0s 2ms/step - loss: 141.5934
Epoch 5/100
11/11 [==============================] - 0s 2ms/step - loss: 176.2341
Epoch 6/100
11/11 [==============================] - 0s 2ms/step - loss: 102.5988
Epoch 7/100
11/11 [==============================] - 0s 2ms/step - loss: 153.2032
Epoch 8/100
11/11 [==============================] - 0s 2ms/step - loss: 126.2234
Epoch 9/100
11/11 [==============================] - 0s 2ms/step - loss: 102.8056
Epoch 10/100
11/11 [==============================] - 0s 2ms/step - loss: 105.3908
Epoch 11/100
11/11 [==============================] - 0s 1ms/step - loss: 124.1387
Epoch 12/100
11/11 [==============================] - 0s 2ms/step - loss: 108.6787
Epoch 13/1

3/3 [==============================] - 0s 2ms/step - loss: 34.0110


34.01097106933594

#### 3) Batch Size

In [48]:
model = Sequential()
model.add(Dense(10, activation = 'relu', input_shape = (n_features,)))
model.add(Dense(8, activation = 'relu'))
model.add(Dense(1))

optimizer = RMSprop(0.1)

model.compile(loss = 'mean_squared_error', optimizer = optimizer)
model.fit(X_train, y_train, epochs = 10, batch_size = 40, verbose = 1)
model.evaluate(X_test, y_test)

Epoch 1/10
9/9 [==============================] - 0s 3ms/step - loss: 8604.3389
Epoch 2/10
9/9 [==============================] - 0s 2ms/step - loss: 293.3398
Epoch 3/10
9/9 [==============================] - 0s 2ms/step - loss: 202.1410
Epoch 4/10
9/9 [==============================] - 0s 2ms/step - loss: 177.6489
Epoch 5/10
9/9 [==============================] - 0s 2ms/step - loss: 171.9786
Epoch 6/10
9/9 [==============================] - 0s 2ms/step - loss: 167.6059
Epoch 7/10
9/9 [==============================] - 0s 3ms/step - loss: 157.3383
Epoch 8/10
9/9 [==============================] - 0s 2ms/step - loss: 267.9005
Epoch 9/10
9/9 [==============================] - 0s 3ms/step - loss: 233.7944
Epoch 10/10
3/3 [==============================] - 0s 0s/step - loss: 79.3241


79.32405853271484

In [49]:
# playing with batch size
learning_rate = 0.01
epochs = 100
batch_size = 40

optimizer = RMSprop(learning_rate)

model.compile(loss = 'mean_squared_error', optimizer = optimizer)
model.fit(X_train, y_train, epochs = epochs, batch_size = batch_size, verbose = 1)
model.evaluate(X_test, y_test)

Epoch 1/100
9/9 [==============================] - 0s 2ms/step - loss: 73.4047
Epoch 2/100
9/9 [==============================] - 0s 2ms/step - loss: 69.0300
Epoch 3/100
9/9 [==============================] - 0s 2ms/step - loss: 65.5986
Epoch 4/100
9/9 [==============================] - 0s 2ms/step - loss: 73.6060
Epoch 5/100
9/9 [==============================] - 0s 2ms/step - loss: 67.1683
Epoch 6/100
9/9 [==============================] - 0s 2ms/step - loss: 65.1247
Epoch 7/100
9/9 [==============================] - 0s 2ms/step - loss: 64.8872
Epoch 8/100
9/9 [==============================] - 0s 2ms/step - loss: 65.8587
Epoch 9/100
9/9 [==============================] - 0s 2ms/step - loss: 63.2532
Epoch 10/100
9/9 [==============================] - 0s 2ms/step - loss: 62.1319
Epoch 11/100
9/9 [==============================] - 0s 2ms/step - loss: 64.3863
Epoch 12/100
9/9 [==============================] - 0s 2ms/step - loss: 66.9339
Epoch 13/100
9/9 [==============================]

27.48406982421875

#### Summary of Hyper-parameter tuning
1)Training loss should steadily decrease, steeply at first, and then more slowly until the slope of the curve reaches or approaches zero.

2)If the training loss does not converge, train for more epochs.

3)If the training loss decreases too slowly, increase the learning rate. Note that setting the learning rate too high may also prevent training loss from converging.

4)If the training loss varies wildly (that is, the training loss jumps around), decrease the learning rate.

5)Lowering the learning rate while increasing the number of epochs or the batch size is often a good combination.

6)Setting the batch size to a very small batch number can also cause instability. First, try large batch size values. Then, decrease the batch size until you see degradation.

7)For real-world datasets consisting of a very large number of examples, the entire dataset might not fit into memory. In such cases, you'll need to reduce the batch size to enable a batch to fit into memory.

#### 5) Making prediction

New test  Data : https://raw.githubusercontent.com/dphi-official/Datasets/master/Boston_Housing/Testing_set_boston.csv

In [50]:
new_test_data = pd.read_csv('https://raw.githubusercontent.com/dphi-official/Datasets/master/Boston_Housing/Testing_set_boston.csv')
new_test_data.head()


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
0  0.09178   0.0   4.05   0.0  0.510  6.416  84.1  2.6463   5.0  296.0   
1  0.05644  40.0   6.41   1.0  0.447  6.758  32.9  4.0776   4.0  254.0   
2  0.10574   0.0  27.74   0.0  0.609  5.983  98.8  1.8681   4.0  711.0   
3  0.09164   0.0  10.81   0.0  0.413  6.065   7.8  5.2873   4.0  305.0   
4  5.09017   0.0  18.10   0.0  0.713  6.297  91.8  2.3682  24.0  666.0   

   PTRATIO       B  LSTAT  
0     16.6  395.50   9.04  
1     17.6  396.90   3.53  
2     20.1  390.11  18.07  
3     19.2  390.91   5.52  
4     20.2  385.09  17.27

In [51]:
model.predict(new_test_data)

4/4 [==============================] - 2s 3ms/step


array([[28.9379   ],
       [34.53398  ],
       [15.273077 ],
       [24.113585 ],
       [18.452108 ],
       [21.701195 ],
       [15.855298 ],
       [ 9.2049465],
       [23.938395 ],
       [20.985891 ],
       [20.474241 ],
       [17.770004 ],
       [ 8.109651 ],
       [19.706024 ],
       [22.712486 ],
       [23.190609 ],
       [17.658155 ],
       [ 8.109651 ],
       [35.944473 ],
       [18.174282 ],
       [25.706903 ],
       [27.710052 ],
       [ 9.454191 ],
       [25.658121 ],
       [16.91954  ],
       [14.770614 ],
       [22.33461  ],
       [ 8.109651 ],
       [20.282507 ],
       [20.065233 ],
       [21.706682 ],
       [27.340572 ],
       [24.301758 ],
       [26.346247 ],
       [10.682259 ],
       [13.731861 ],
       [32.49585  ],
       [23.21467  ],
       [19.836119 ],
       [21.010078 ],
       [11.394365 ],
       [31.7815   ],
       [36.87091  ],
       [17.534233 ],
       [26.876036 ],
       [18.925808 ],
       [14.695684 ],
       [22.57